### PR TITLE
GH-3474: add random seed parameter to dataset splitting and downsampl…

### PR DESCRIPTION
**Pull request by MattGPT-ai**
https://github.com/flairNLP/flair/issues/3474

Adds random seed parameters to dataset splitting and downsampling functions

### CODE REVIEW

**Review by helpmefindaname — CHANGES_REQUESTED** [review]
Hi @MattGPT-ai
thank you for your contribution.
The changes look good to me, I just requests a small change. Besides that, I think this is ready to merge

File: flair/data.py [hunk 1]
```diff
@@ -1879,11 +1882,21 @@ def iob2(tags):
     return True
 
 
-def randomly_split_into_two_datasets(dataset, length_of_first):
+def randomly_split_into_two_datasets(
+    dataset: Dataset, length_of_first: int, random_seed: Optional[int] = None
+) -> tuple[Subset, Subset]:
```
Comment (helpmefindaname) [inline flair/data.py:1887]: Please use `Tuple` (from typing import Tuple) to allow compability with older python versions

**Comment by MattGPT-ai** [comment]
Thanks for your review! Pushed the changes you requested

**Review by helpmefindaname — APPROVED** [review]

**Comment by MattGPT-ai** [comment]
Thanks for the approval! How can we trigger the checks to run so that it can be merged?

File: flair/data.py [hunk 2]
```diff
@@ -1879,11 +1882,21 @@ def iob2(tags):
...
+def randomly_split_into_two_datasets(
+    dataset: Dataset, length_of_first: int, random_seed: Optional[int] = None
+) -> Tuple[Subset, Subset]:
+    """Shuffles a dataset and splits into two subsets.
+
+    The length of the first is specified and the remaining samples go into the second subset.
+    """
     import random
 
     indices = list(range(len(dataset)))
```
Comment (alanakbik) [inline flair/data.py:1894]:
```suggestion
    indices = list(range(_len_dataset(dataset)))
```

**Review by alanakbik — CHANGES_REQUESTED** [review]
Thanks for the PR! Looks good, but there is a small mypy error (I added a suggestion to fix).

**Comment by MattGPT-ai** [comment]
Applied the change. I had a question, I saw that step is necessary because Dataset doesn't have `__len__` implemented. Is there a way to implement `__len__`, even if it just calls `_len_dataset`? Does constructing the `DatasetLoader` class take any non-negligible time? I actually see the same warning for `return len(loader)` in my PyCharm that it "Expected type Sized, got DataLoader instead", and a similar warning for `_iter_dataset`

**Comment by alanakbik** [comment]
Thanks! I agree that it would be better to implement `__len__` rather than having to call `_len_dataset`. 

The problem is that `Dataset` is a PyTorch class that does not have `__len__` or `__iter__` implemented, so we can't add them here. Alternatively, we might change the code that instead of `Dataset`, everything is based on either PyTorch's `IterDataset` or our own `FlairDataset`.

**Review by alanakbik — APPROVED** [review]

**Comment by alanakbik** [comment]
@MattGPT-ai thanks a lot for adding this!
